cpp/src/arrow/compute/kernels/comparator.cc

```diff
@@ -0,0 +1,197 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/kernels/comparator.h"
+
+#include "arrow/compute/context.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/filter.h"
+#include "arrow/compute/kernels/util-internal.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/util/logging.h"
+
+namespace arrow {
+
+namespace compute {
+
+class FunctionContext;
+struct Datum;
+
+static inline Status CopyOrTransferBitmap(const Array& array, MemoryPool* pool,
+                                          std::shared_ptr<Buffer>* bitmap) {
+  auto offset = array.offset();
+  auto length = array.length();
+  auto null_bitmap = array.null_bitmap();
+
+  if (null_bitmap == nullptr || offset == 0 || length == 0) {
+    *bitmap = null_bitmap;
```

Comment:
nit
```suggestion
    *bitmap = nullptr;
```

Reply:
No, if `offset = 0`, it preserves the original buffer in a zero-copy fashion.

Reply:
I'll make it explicit in another branch.